bentoml/_internal/utils/telemetry.py

```diff
@@ -0,0 +1,65 @@
+# type: ignore[reportMissingTypeStubs]
+import typing as t
+from typing import TYPE_CHECKING
+
+from opentelemetry.trace import get_current_span
+from opentelemetry.sdk.trace.sampling import Decision
+from opentelemetry.sdk.trace.sampling import ParentBased
+from opentelemetry.sdk.trace.sampling import SamplingResult
+from opentelemetry.sdk.trace.sampling import TraceIdRatioBased
+
+if TYPE_CHECKING:
+    from opentelemetry.trace import Link
+    from opentelemetry.trace import Context
+    from opentelemetry.trace import SpanKind
+    from opentelemetry.trace import TraceState
+    from opentelemetry.util.types import Attributes
+
+
+def _get_parent_trace_state(parent_context: "Context") -> t.Optional["TraceState"]:
+    parent_span_context = get_current_span(parent_context).get_span_context()
+    if parent_span_context is None or not parent_span_context.is_valid:
+        return None
+    return parent_span_context.trace_state
+
+
+class TraceIdRatioBasedAlwaysRecording(TraceIdRatioBased):
+    """
+    A trace Sampler that:
+    * always recording (so that we can get the trace_id)
+    * respect the parent's trace_state
+    * ratio sampling
+    """
```

Comment:
I need to think about this more but we might be more fitting to use `ParentBased` instead of `TraceIdRatioBased` here. If BentoML is a part of a bigger trace, the sampling decision might already be made at one of the earlier services. BentoML should respect that trace decision. If BentoML API server is the root, we can then define and leverage our own sample rate. Let's discuss more tomorrow.

Reply:
Actually we are using ParrentBasedAndTraceIdRatioBasedAlwaysRecording